flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryService.java

```diff
@@ -0,0 +1,103 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.exceptionhistory;
+
+import org.apache.flink.api.common.JobStatus;
+import org.apache.flink.runtime.executiongraph.JobStatusProvider;
+import org.apache.flink.runtime.executiongraph.failover.flip1.FailureHandlingResult;
+import org.apache.flink.runtime.scheduler.ExceptionHistoryEntry;
+import org.apache.flink.runtime.util.BoundedFIFOQueue;
+
+import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** {@code ExceptionHistoryService} collects failure causes. */
+public class ExceptionHistoryService {
+
+    private final Logger log;
+
+    private final JobStatusProvider jobStatusProvider;
+    private final ExceptionHistoryEntryFactory exceptionHistoryEntryFactory;
+
+    private final BoundedFIFOQueue<ExceptionHistoryEntry> entries;
+
+    public ExceptionHistoryService(
+            JobStatusProvider jobStatusProvider,
+            ExceptionHistoryEntryFactory exceptionHistoryEntryFactory,
+            int maxQueueSize,
+            Logger log) {
+        this(
+                jobStatusProvider,
+                exceptionHistoryEntryFactory,
+                new BoundedFIFOQueue<>(maxQueueSize),
+                log);
+    }
+
+    private ExceptionHistoryService(
+            JobStatusProvider jobStatusProvider,
+            ExceptionHistoryEntryFactory exceptionHistoryEntryFactory,
+            BoundedFIFOQueue<ExceptionHistoryEntry> entries,
+            Logger log) {
+        this.jobStatusProvider = jobStatusProvider;
+        this.exceptionHistoryEntryFactory = exceptionHistoryEntryFactory;
+        this.entries = entries;
+        this.log = log;
+    }
+
+    public void registerGlobalFailure(Throwable throwable) {
+        registerGlobalFailure(throwable, jobStatusProvider.getStatusTimestamp(JobStatus.FAILED));
+    }
+
+    public void registerGlobalFailure(@Nullable Throwable throwable, long timestamp) {
+        entries.add(ExceptionHistoryEntry.fromGlobalFailure(throwable, timestamp));
+        log.debug("Archive global failure.", throwable);
+    }
+
+    public void registerLocalFailure(FailureHandlingResult failureHandlingResult) {
```

Comment:
I'm not convinced that passing in an AccessExecution makes testing much more difficult. And then we could forego this whole execption entry factory business.